staging/src/k8s.io/apiserver/pkg/cel/openapi/resolver.go

```diff
@@ -0,0 +1,152 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+type SchemaResolver interface {
+	ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error)
+}
+
+var ErrSchemaNotFound = fmt.Errorf("schema not found")
+
+// ClientDiscoveryResolver uses client-go discovery to resolve schemas at run time.
+type ClientDiscoveryResolver struct {
+	Discovery discovery.DiscoveryInterface
+}
+
+var _ SchemaResolver = (*ClientDiscoveryResolver)(nil)
+
+func (r *ClientDiscoveryResolver) ResolveSchema(gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	p, err := r.Discovery.OpenAPIV3().Paths()
+	if err != nil {
+		return nil, err
+	}
+	resourcePath := resourcePathFromGV(gvk.GroupVersion())
+	c, ok := p[resourcePath]
+	if !ok {
+		return nil, fmt.Errorf("cannot resolve group version %q: %w", gvk.GroupVersion(), ErrSchemaNotFound)
+	}
+	b, err := c.Schema(runtime.ContentTypeJSON)
+	if err != nil {
+		return nil, err
+	}
+	resp := new(schemaResponse)
+	err = json.Unmarshal(b, resp)
+	if err != nil {
+		return nil, err
+	}
+	s, err := resolveType(resp, gvk)
+	if err != nil {
+		return nil, err
+	}
+	err = r.populateRefs(resp, s)
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func (r *ClientDiscoveryResolver) populateRefs(resp *schemaResponse, schema *spec.Schema) error {
+	ref, isRef := refOf(schema)
+	if isRef {
+		// replace the whole schema with the referred one.
+		resolved, ok := resp.Components.Schemas[strings.TrimPrefix(ref, refPrefix)]
+		if !ok {
+			return fmt.Errorf("internal error: cannot resolve Ref %q: %w", ref, ErrSchemaNotFound)
+		}
+		*schema = *resolved
+	}
+	// schema is an object, populate its properties and additionalProperties
+	for name, prop := range schema.Properties {
+		err := r.populateRefs(resp, &prop)
+		if err != nil {
+			return err
+		}
+		schema.Properties[name] = prop
+	}
+	if schema.AdditionalProperties != nil && schema.AdditionalProperties.Schema != nil {
+		err := r.populateRefs(resp, schema.AdditionalProperties.Schema)
+		if err != nil {
+			return err
+		}
+	}
+	// schema is a list, populate its items
+	if schema.Items != nil && schema.Items.Schema != nil {
+		err := r.populateRefs(resp, schema.Items.Schema)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func refOf(schema *spec.Schema) (string, bool) {
+	if schema.Ref.GetURL() != nil {
+		return schema.Ref.String(), true
+	}
+	for _, allOf := range schema.AllOf {
+		if allOf.Ref.GetURL() != nil {
+			return allOf.Ref.String(), true
+		}
+	}
+	return "", false
+}
+
+func resolveType(resp *schemaResponse, gvk schema.GroupVersionKind) (*spec.Schema, error) {
+	for _, s := range resp.Components.Schemas {
```

Comment:
Can `resp` or `resp.Components` ever be nil?

Reply:
No because `resp` is unmarshalled from JSON, and `Components` is not a pointer. If the returned structure of OpenAPI v3 endpoint ever changes, worst case would be a no-op loop.